### PR TITLE
Rewrite tests for ``split_non_commuting``

### PR DESCRIPTION
**Pull request by astralcai**
This is an attempt to refactor the tests for one of the transforms, hoping to set a standard for future transform tests. For each transform, we should have:

- Unit tests for basic functionalities of the transform:
  - Does the transform produce the correct tapes?
  - Does the processing function return the correct results (numerical correctness tested here)?
- Tests for integration of the transform with a QNode:
  - Tests different combinations of parameter broadcasting and shots/shot-vectors
  - Mainly tests that the circuit executes without error and returns results with the correct shapes.
- Tests for the differentiability of QNodes with this transform applied:
  - Tests different interfaces
  - Tests numerical correctness of the gradients

[sc-72000]
[sc-72519]

### CODE REVIEW

File: pennylane/ops/qubit/non_parametric_ops.py [hunk 1]
```diff
@@ -57,6 +57,9 @@ class Hadamard(Observable, Operation):
 
     _queue_category = "_ops"
 
+    def __init__(self, wires: Optional[WiresLike] = None, id: Optional[str] = None):
```
Comment (astralcai) [inline pennylane/ops/qubit/non_parametric_ops.py:60]: This is such that an IDE does not complain about the `0` in `qml.H(0)` not being "`TensorLike`"

File: pennylane/transforms/split_non_commuting.py [hunk 2]
```diff
@@ -288,26 +288,15 @@ def circuit(x):
...
-            isinstance(m, ExpectationMP) and isinstance(m.obs, LinearCombination)
-            for m in tape.measurements
-        )
-        or any(
-            m.obs is not None and not qml.pauli.is_pauli_word(m.obs) for m in single_term_obs_mps
-        )
+    if grouping_strategy == "wires" or any(
+        m.obs is not None and not qml.pauli.is_pauli_word(m.obs) for m in single_term_obs_mps
     ):
-        # This is a loose check to see whether wires grouping or qwc grouping should be used,
```
Comment (astralcai) [inline pennylane/transforms/split_non_commuting.py:302]: The premise of the old logic here no longer holds, the graph-based grouping method consistently outperforms wire-based grouping after a refactor of the grouping module completed a few months ago, so there is no reason to use wire-based grouping as a default anymore.

File: pennylane/ops/op_math/sum.py [hunk 3]
```diff
@@ -525,8 +525,9 @@ def compute_grouping(self, grouping_type="qwc", method="lf"):
         _, ops = self.terms()
 
         with qml.QueuingManager.stop_recording():
-            op_groups = qml.pauli.group_observables(ops, grouping_type=grouping_type, method=method)
-        self._grouping_indices = tuple(tuple(ops.index(o) for o in group) for group in op_groups)
+            self._grouping_indices = qml.pauli.compute_partition_indices(
```
Comment (albi3ro) [inline pennylane/ops/op_math/sum.py:528]: 😬 that was a fairly bad oversight... glad we caught it though?

Comment (albi3ro) [inline pennylane/ops/op_math/sum.py:528]: Was this causing a test to fail?

Comment (albi3ro) [reply pennylane/ops/op_math/sum.py:528]: Also note that there's a technical roadmap task to make this change.

File: pennylane/transforms/split_non_commuting.py (hunk 2 above)
Comment (albi3ro) [reply pennylane/transforms/split_non_commuting.py:302]: Should we make this change in an independent PR?  Given the premise for this PR is "tests" and not "performance improvement for grouping".

File: pennylane/ops/op_math/sum.py (hunk 3 above)
Comment (astralcai) [reply pennylane/ops/op_math/sum.py:528]: > Was this causing a test to fail?
No, I just stumbled upon it and it felt like a straightforward enough change.

File: pennylane/transforms/split_non_commuting.py (hunk 2 above)
Comment (astralcai) [reply pennylane/transforms/split_non_commuting.py:302]: The thing is, if we do not make the change here, we would have to change the tests to expect the old behavior, and when we do make the change, update the tests back to expect the new behavior.

File: tests/transforms/test_split_non_commuting.py [hunk 4]
```diff
@@ -95,7 +92,7 @@ def complex_qwc_processing_fn(results):
     return (
         group0[0],
         0.5 * group1[0],
-        group0[0] + group1[1] + 2.0 * group2[0] + 1.0,
+        group0[0] + group1[1] + 2.0 * group2 + 1.0,
```
Comment (mlxd) [inline tests/transforms/test_split_non_commuting.py:95]: Is there any harm in `group2[0]` to match the others, or is it outright wrong?

File: tests/transforms/test_split_non_commuting.py [hunk 5]
```diff
@@ -122,765 +119,580 @@ def complex_wires_processing_fn(results):
...
-        "make_H",
-        [
-            lambda coeffs, obs: qml.Hamiltonian(coeffs, obs),
-            lambda coeffs, obs: qml.sum(*(qml.s_prod(c, o) for c, o in zip(coeffs, obs))),
-        ],
-    )
-    def test_number_of_tapes_single_hamiltonian(self, grouping_strategy, n_tapes, make_H):
-        """Tests that the correct number of tapes is returned for a single Hamiltonian"""
+    group0, group1, group2, group3 = results
+    return group0[0], 0.5 * group1[0], group0[0] + group2 + 2.0 * group0[1] + group3, group1[1]
```
Comment (mlxd) [inline tests/transforms/test_split_non_commuting.py:143]:
```suggestion
    res = (group0[0], 0.5 * group1[0], group0[0] + group2 + 2.0 * group0[1] + group3, group1[1])
    return res
```

Could we do something like this here and above? It makes step-wise debugging easier to inspect.

File: tests/transforms/test_split_non_commuting.py [hunk 6]
```diff
@@ -122,765 +119,580 @@ def complex_wires_processing_fn(results):
...
-        tapes, fn = split_non_commuting(tape, grouping_strategy="qwc")
-        for actual_tape, expected_tape in zip(tapes, expected_tapes_qwc_grouping):
-            qml.assert_equal(actual_tape, expected_tape)
-        assert qml.math.allclose(fn([[0.1, 0.2], [0.3, 0.4, 0.5]]), [0.01, 0.06, 0.12, 0.08, 0.25])
+    def test_single_hamiltonian_precomputed_grouping(
+        self, H, grouping_indices, results, expected_result
+    ):
+        """Tests that precomputed grouping of a single Hamiltonian is used."""
+
+        H._grouping_indices = grouping_indices  # pylint: disable=protected-access
```
Comment (mlxd) [inline tests/transforms/test_split_non_commuting.py:195]: Is there an alternative way to access these values?

File: tests/transforms/test_split_non_commuting.py [hunk 7]
```diff
@@ -122,765 +119,580 @@ def complex_wires_processing_fn(results):
...
+        initial_tape = qml.tape.QuantumScript([qml.X(0)], [qml.expval(H)], shots=100)
+        tapes, fn = split_non_commuting(initial_tape)
+        assert len(tapes) == len(grouping_indices)
+
+        for group_idx, group in enumerate(grouping_indices):
+            assert tapes[group_idx].measurements == [
+                qml.expval(single_term_obs_list[i])
+                for i in group
+                if not isinstance(single_term_obs_list[i], qml.Identity)
+            ]
```
Comment (mlxd) [inline tests/transforms/test_split_non_commuting.py:205]: Could this be split over a few statements? It's quite hard to parse as a single assertion.
We can create the list in a separate line (or several) then validate the assert

File: tests/transforms/test_split_non_commuting.py [hunk 8]
```diff
@@ -122,765 +119,580 @@ def complex_wires_processing_fn(results):
...
+        initial_tape = qml.tape.QuantumScript([qml.X(0)], [qml.expval(H)], shots=100)
+        tapes, fn = split_non_commuting(initial_tape, grouping_strategy=grouping_strategy)
+        assert H.grouping_indices == ((0, 3, 5), (1, 2, 4))
+
+        for group_idx, group in enumerate(H.grouping_indices):
+            assert tapes[group_idx].measurements == [
+                qml.expval(single_term_obs_list[i])
+                for i in group
+                if not isinstance(single_term_obs_list[i], qml.Identity)
+            ]
```
Comment (mlxd) [inline tests/transforms/test_split_non_commuting.py:231]: Same state here as above --- splitting this would make it easier to understand.

File: tests/transforms/test_split_non_commuting.py [hunk 9]
```diff
@@ -122,765 +119,580 @@ def complex_wires_processing_fn(results):
...
+        assert tapes[0].operations == [qml.X(0)]
+        assert qml.math.allclose(fn([0.1]), [0.1, 0.05])
+
+    def test_mix_measurement_types_qwc(self):
+        """Tests multiple measurement types can be handled by qwc grouping"""
+
+        initial_tape = qml.tape.QuantumScript(
+            [qml.X(0)],
+            measurements=[
+                # [Z(0), X(0), Z(0), X(0), Y(1), Z(1), Z(0) @ Z(1), Z(1), Z(0)]
```
Comment (mlxd) [inline tests/transforms/test_split_non_commuting.py:412]: Is this relevant to the statements below? Better to make it clear if so

File: tests/transforms/test_split_non_commuting.py [hunk 10]
```diff
@@ -122,765 +119,580 @@ def complex_wires_processing_fn(results):
...
-    )
-    def test_tape_with_non_pauli_obs(self, non_pauli_obs):
-        """Tests that the tape is split correctly when containing non-Pauli observables"""
-
-        obs_list = single_term_obs_list + non_pauli_obs
+        )
 
-        measurements = [
-            qml.expval(c * o) for c, o in zip([0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7], obs_list)
+        # ((0, 2, 5, 6, 7, 8), (1, 3, 4))
```
Comment (mlxd) [inline tests/transforms/test_split_non_commuting.py:425]: Same as above -- is this relevant to the statements below?

File: tests/transforms/test_split_non_commuting.py [hunk 11]
```diff
@@ -122,765 +119,580 @@ def complex_wires_processing_fn(results):
...
+            0.2,
+            np.array([0.5, 0.5]),
+            {"1": 2},
+            np.array([1.0, 1.0]),
+            np.array([1.0, 0.0, 0.0, 0.0]),
+            np.array([0, 0]),
+            {"0": 2},
+        )
+        assert len(results) == len(expected)
+        for res, exp in zip(results, expected):
```
Comment (mlxd) [inline tests/transforms/test_split_non_commuting.py:466]: I'd generally discourage shadowing names from math operations --- `exp` would be best named something else

File: tests/transforms/test_split_non_commuting.py [hunk 12]
```diff
@@ -122,765 +119,580 @@ def complex_wires_processing_fn(results):
...
+        @partial(split_non_commuting, grouping_strategy=grouping_strategy)
         @qml.qnode(dev)
         def circuit(angles):
             qml.RX(angles[0], wires=0)
-            qml.RY(angles[1], wires=0)
-            qml.RX(angles[0], wires=1)
             qml.RY(angles[1], wires=1)
-            return [qml.expval(obs) for obs in obs_list]
+            qml.RZ(angles[2], wires=2)
+            return qml.expval(qml.Hamiltonian(coeffs, obs))
```
Comment (mlxd) [inline tests/transforms/test_split_non_commuting.py:607]: Do we need an execution for this validation?

File: tests/transforms/test_split_non_commuting.py [hunk 13]
```diff
@@ -122,765 +119,580 @@ def complex_wires_processing_fn(results):
...
-        """Tests that a QNode with mixed measurement types is executed correctly"""
+    def test_general_circuits(self, grouping_strategy, shots, params):
+        """Tests executing a QNode with different grouping strategies on a typical circuit."""
 
-        dev = qml.device("default.qubit", wires=2, shots=shots, seed=seed)
-
-        obs_list = complex_obs_list
+        dev = qml.device("default.qubit", wires=3, shots=shots)
 
+        @partial(split_non_commuting, grouping_strategy=grouping_strategy)
```
Comment (mlxd) [inline tests/transforms/test_split_non_commuting.py:630]: Same as above --- do we need an execution to know this is correct?

File: tests/transforms/test_split_non_commuting.py [hunk 14]
```diff
@@ -122,765 +119,580 @@ def complex_wires_processing_fn(results):
...
 
-        with _dev.tracker:
-            res = circuit()
+            assert len(_result) == _expected_shape[0]
+            for res in _result:
+                _recursively_check_shape(res, _expected_shape[1:])
 
-        assert _dev.tracker.totals == {}
-        assert qml.math.allclose(res, [1.5, 2.5])
+        _recursively_check_shape(result, expected_shape)
```
Comment (mlxd) [inline tests/transforms/test_split_non_commuting.py:672]: Any worry about recursive calls with this?

File: tests/transforms/test_split_non_commuting.py [hunk 15]
```diff
@@ -122,765 +119,580 @@ def complex_wires_processing_fn(results):
...
+    @pytest.mark.parametrize("grouping_strategy", [None, "qwc", "wires"])
+    @pytest.mark.parametrize(
+        "obs, expected_result",
+        [
+            ([qml.Hamiltonian([1.5, 2.5], [qml.I(0), qml.I(1)])], 4.0),
+            ([1.5 * qml.I(), 2.5 * qml.I()], [1.5, 2.5]),
+        ],
+    )
+    def test_only_constant_offset(self, grouping_strategy, obs, expected_result):
+        """Tests that split_non_commuting can handle a circuit only measuring Identity."""
```
Comment (mlxd) [inline tests/transforms/test_split_non_commuting.py:683]: Do we need an execution for this test?

**Review by mlxd — COMMENTED** [review]
Thanks @astralcai 

A few questions from my side:

File: tests/transforms/test_split_non_commuting.py (hunk 4 above)
Comment (astralcai) [reply tests/transforms/test_split_non_commuting.py:95]: It's outright wrong, because `group2` corresponds to a tape that only has one measurement. This dimension should've been squeezed out.

File: tests/transforms/test_split_non_commuting.py (hunk 12 above)
Comment (astralcai) [reply tests/transforms/test_split_non_commuting.py:607]: These are integration tests of applying the transform on a QNode and making sure that it executes correctly so I would say yes?

File: tests/transforms/test_split_non_commuting.py (hunk 14 above)
Comment (astralcai) [reply tests/transforms/test_split_non_commuting.py:672]: Nope, the `expected_shape` that gets passed down the chain of recursion is shrinking in length.

File: tests/transforms/test_split_non_commuting.py (hunk 15 above)
Comment (astralcai) [reply tests/transforms/test_split_non_commuting.py:683]: Yes, because this is a special case that resulted in no execution at all (zero tapes), and we want to make sure that it does not error out or something.

File: tests/transforms/test_split_non_commuting.py (hunk 9 above)
Comment (astralcai) [reply tests/transforms/test_split_non_commuting.py:414]: Added a comment.

File: tests/transforms/test_split_non_commuting.py (hunk 6 above)
Comment (mlxd) [reply tests/transforms/test_split_non_commuting.py]: This was resolved without comment. Same question as above.

File: tests/transforms/test_split_non_commuting.py (hunk 7 above)
Comment (mlxd) [reply tests/transforms/test_split_non_commuting.py]: This was resolved without comment.

File: tests/transforms/test_split_non_commuting.py (hunk 8 above)
Comment (mlxd) [reply tests/transforms/test_split_non_commuting.py]: Resolved without comment.

**Review by mlxd — CHANGES_REQUESTED** [review]
Many previous questions resolved without comment. Same questions/prompts apply.

File: tests/transforms/test_split_non_commuting.py (hunk 6 above)
Comment (astralcai) [reply tests/transforms/test_split_non_commuting.py]: Sorry about that. I've updated the code to use the public `grouping_indices` instead, that's why I marked it as resolved.

File: tests/transforms/test_split_non_commuting.py (hunk 7 above)
Comment (astralcai) [reply tests/transforms/test_split_non_commuting.py]: I've split this into multiple lines.

File: tests/transforms/test_split_non_commuting.py (hunk 8 above)
Comment (astralcai) [reply tests/transforms/test_split_non_commuting.py]: I've split this into multiple lines. It appears that if you directly click the link to the code from the comment, it takes you to an old version of the code. You can see the updated lines if you manually navigate to its location.

File: tests/transforms/test_split_non_commuting.py (hunk 10 above)
Comment (astralcai) [reply tests/transforms/test_split_non_commuting.py:428]: Also added a comment

File: tests/transforms/test_split_non_commuting.py (hunk 13 above)
Comment (astralcai) [reply tests/transforms/test_split_non_commuting.py:633]: I feel that some integration testing with execution is needed, just to make sure that everything works, especially with different return shapes and types associated with parameter broadcasting and shot vectors.

File: pennylane/ops/qubit/non_parametric_ops.py (hunk 1 above)
Comment (albi3ro) [reply pennylane/ops/qubit/non_parametric_ops.py:60]:
```suggestion
    def __init__(self, wires: WiresLike, id: Optional[str] = None):
```
Though the issue with built-ins and `TensorLike` is a deeper issue we need to address at some point.

File: tests/transforms/test_split_non_commuting.py (hunk 4 above)
Comment (albi3ro) [reply tests/transforms/test_split_non_commuting.py:95]: We squeeze out the single dimension of the results when there is a single measurement process.

File: tests/transforms/test_split_non_commuting.py (hunk 6 above)
Comment (mlxd) [reply tests/transforms/test_split_non_commuting.py]: All good, thanks. Just making sure things don't go unnoticed :)

File: tests/transforms/test_split_non_commuting.py (hunk 12 above)
Comment (mlxd) [reply tests/transforms/test_split_non_commuting.py:618]: I guess this is something we can mock out the device with, as it doesn't need the device end-to-end, just the QN integration, right? Though, I'd fine with making a note/follow up story to do this and proceed as is for now.

File: tests/transforms/test_split_non_commuting.py (hunk 13 above)
Comment (mlxd) [reply tests/transforms/test_split_non_commuting.py:641]: Same as above --- as this is a system (end-to-end) rather than integration test, we can come back later and mock out the device execution to see the interplay with the qnode only.

File: tests/transforms/test_split_non_commuting.py (hunk 12 above)
Comment (astralcai) [reply tests/transforms/test_split_non_commuting.py:618]: The test was marked `pytest.mark.system`. I guess I shouldn't have named it `TestQNodeIntegration`. I meant to test the full execution pipeline, because we want to be testing with real result batches returned from device execution of the tapes that we generated. I wrote the original test, and I believe these types of tests did reveal certain edge cases that wasn't considered before.

Comment (mlxd) [reply tests/transforms/test_split_non_commuting.py:618]: Let's make note of this separation to happen as a future TODO, and plan to reduce these testing overheads. I still see system level tests here as non essential, and to be brought to unit/integration level.
If you are happy with that assessment and we can capture those items for future work, I'm happy to approve as is.

Comment (astralcai) [reply tests/transforms/test_split_non_commuting.py:618]: I think it depends on the transform. For some transforms where nothing complicated happens in post-processing, this is non-essential, but the `split_non_commuting` contains complicated post-processing of the execution results. It is technically possible to test the post processing transform with mocked device execution, but given the complexity of the return type specification (5 dimensions, each of them may or may not have been squeezed out in different scenarios), I'm not confident that the mock is able to exactly mimic the behaviour of actual execution, and it's probably not worth it to do so. For these tests I have set the number of shots to `10` to avoid too much overhead, since we're not testing for numerical correctness here, just that the circuit executes without error and the returned values have the correct shapes.

Comment (mlxd) [reply tests/transforms/test_split_non_commuting.py:618]: I think we need to disagree and commit on this @astralcai :) 
Let's track this item for future work so we can get this PR through.

Comment (astralcai) [reply tests/transforms/test_split_non_commuting.py:618]: Sounds good. I've created a story on SC and tagged you in it. @mlxd

Comment (mlxd) [reply tests/transforms/test_split_non_commuting.py:618]: Awesome, TY.

**Review by mlxd — APPROVED** [review]

**Review by albi3ro — APPROVED** [review]
